File: MATLAB/+plt/plotSusp.m
function plotSusp(P,linefrmt,rTyre)

%upper wishbone
b1 = [P(:,1), P(:,2), P(:,3)];
plt.plotBody(b1, linefrmt, 'r'); hold on

%lower wishbone
b2 = [P(:,4), P(:,5), P(:,6)];
plt.plotBody(b2,linefrmt,'g');

%upright axis
b2 = [P(:,3), P(:,6), P(:,10)];
plt.plotBody(b2,linefrmt,'k');

%damper axis
b3 = [P(:,7), P(:,8)];
plt.plotBody(b3,linefrmt,'b');

%spring axis
b5 = [P(:,11), P(:,12)];
plt.plotBody(b5,linefrmt,'b');

%track-rod
b4 = [P(:,9), P(:,10)];
plt.plotBody(b4,linefrmt,'c');


%wheel axis
b6 = [P(:,13), P(:,14)];
plt.plotBody(b6,linefrmt,'k');

% aCamber0 = -90+(acosd(dot(HP(:,14)-HP(:,13), [0;0;1])/norm(HP(:,14)-HP(:,13))));
% aToe0 = 90-(acosd(dot(HP(:,14)-HP(:,13), [1;0;0])/norm(HP(:,14)-HP(:,13))));
% plt.plotTyre(P(:,14),rTyre,aCamber0,aToe0);

% bodies(:,:,1) = [P1;P2];            %chassis
% bodies(:,:,2) = [P1;P2];            %upper link (wishbone)
% bodies(:,:,3) = [P1;P2];            %lower link (wishbone)
% bodies(:,:,4) = [P1;P2];            %oubj and olbj upright
% bodies(:,:,5) = [P1;P2];            %chassis
% bodies(:,:,6) = [P1;P2];            %chassis
% 
% l0 = [P1;P2];           
% l1 = [P1;P3];           %upper link (wishbone)
% l2 = [P2;P4];           %lower link (wishbone)
% l3 = [P3;P4];           %oubj and olbj upright
% l4 = [P4;P5];           %olbj and contact patch
% l5 = [P3;P5];           %oubj and contact patch
% l6 = [P5;P6];           %wheel/tyre centerline
% 
% 
% 
% %Links' distances (euclidian)
% d1 = norm(P1-P3);
% d2 = norm(P2-P4);
% d3 = norm(P3-P4);
% d4 = norm(P4-P5);
% d5 = norm(P3-P5);
% 
% %Visualization
% f = figure(1);
% plot(l1(:,1), l1(:,2),'o-b','linewidth',3,'MarkerSize',6); hold on;
% plot(l2(:,1), l2(:,2),'o-g','linewidth',3,'MarkerSize',6);
% plot(l3(:,1), l3(:,2),'o-r','linewidth',3,'MarkerSize',6);
% plot(l4(:,1), l4(:,2),'o-r','linewidth',3,'MarkerSize',6);
% plot(l5(:,1), l5(:,2),'o-r','linewidth',3,'MarkerSize',6);
% plot(l0(:,1), l0(:,2),'o-k','linewidth',3,'MarkerSize',6);
% plot(l6(:,1), l6(:,2),'--k','linewidth',0.5);
% 
% text(P1(1)-1,P1(2)+1,'UIBJ');
% text(P2(1)-1,P2(2)-1,'LIBJ');
% text(P3(1)-1,P3(2)+1,'UOBJ');
% text(P4(1)-1,P4(2)-1,'LOBJ');
% text(P5(1)+1,P5(2)+1,'CP');
% 
% %Instant Center
% m1 = (P3(2)-P1(2))/(P3(1)-P1(1));
% m2 = (P4(2)-P2(2))/(P4(1)-P2(1));
% b1 = P3(2)-m1*P3(1);
% b2 = P4(2)-m2*P4(1);
% x = (b2-b1)/(m1+m2);
% y = m1*x+b1;
% 
% IC = [x, y];
% plot(IC(1), IC(2),'*K'); plot([IC(1), P1(1), P2(1), IC(1)],[IC(2), P1(2), P2(2), IC(2)],'--K');
% text(IC(1)-1,IC(2)+1,'IC');
% %Roll Center
% RCx = 0;                    %Assuming symmetric suspension geometry
% if ~isfinite(IC(1))
%     RCy = 0;
% else
%     mr = (P5(2)-y)/(P5(1)-x);
%     RCy = P5(2) - mr*P5(1);
% end
% 
% RC = [RCx, RCy];
% plot(RC(1),RC(2),'oK','MarkerSize',10);
% plot([IC(1),P5(1)], [IC(2),P5(2)],'--K');
% text(RC(1),RC(2)+1,'RC');
% 
% axis([IC(1)-10, P5(1)+10, P5(2)-10, P1(2) + 20]);
% 
% %Basic metrics
% trkWidth = 2*P5(1);
% aKingpin = atand((P3(1)-P4(1))/(P3(2)-P4(2)));
% 
% if P3(1) == P4(1)
%     dScrub = P5(1)-P3(1);
% else
%     dScrub = P5(1) + (P3(2) - ((P3(2)-P4(2))/(P3(1)-P4(1)))*P3(1))/((P3(2)-P4(2))/(P3(1)-P4(1)));
% end
% 
% display("Track width: " + num2str(trkWidth));
% display("Kingpin angle: " + num2str(aKingpin));
% display("Scrub radius: " + num2str(dScrub));
% display("Virtual swing arm length: " + num2str(P5(1)-IC(1)));
% display("Roll center height: " + num2str(RCy));
% display("Camber Gain Rate: " + num2str(atand(1/(P5(1)-IC(1)))));

File: Python/plot_susp.py
import matplotlib.pyplot as plt
import numpy as np
import math


def plot_body(ax, body, linefrmt, clr, close=1):
    if close == 1 and body.shape[1] > 2:
        body = np.append(body, body[:, 0].reshape(3, 1), axis=1)

    ax.plot(body[0, :], body[1, :], body[2, :], linefrmt, color=clr)


def plot_susp(HP, virtual=False, mirror=False):


    fig = plt.figure()
    ax = fig.add_subplot(projection='3d')
    ax.view_init(0, 0, 0)
    ax.set_xlabel('X Axis')
    ax.set_ylabel('Y Axis')
    ax.set_zlabel('Z Axis')

    if mirror is True:
        i = [1, -1]
    else:
        i = [1]

    for kx in i:

        HP[1, :] = kx * HP[1, :]

        # Upper A-Arm
        b1 = HP[:, :3]
        plot_body(ax, b1, 'o-', 'red')

        # Lower A-Arm
        b2 = HP[:, 3:6]
        plot_body(ax, b2, 'o-', 'green')

        # Upright
        b3 = HP[:, [2, 5, 9]]
        plot_body(ax, b3, 'o-', 'black')

        # Tie Rod
        b4 = HP[:, 8:10]
        plot_body(ax, b4, 'o-', 'cyan')

        # Spring/Damper Axis
        b5 = HP[:, 6:8]
        plot_body(ax, b5, 'o-', 'blue')

        # Wheel Axis
        b6 = HP[:, 12:14]
        plot_body(ax, b6, 'o-', 'grey')

        # Tyre Contact Patch
        ax.plot(HP[0, 14], HP[1, 14], HP[2, 14], '*', color='black')

        if virtual is True:
            plot_virtual(ax, HP)

    ax.set_aspect('equal', adjustable='box')
    plt.show()


def plot_virtual(ax, HP, color='silver'):

    VP = get_virtual(HP)

    ax.plot([0,0,0], VP[0, [0,2,4]], VP[1, [0,2,4]], '--', color=color)
    ax.plot([0,0,0], VP[0, [1, 3, 4]], VP[1, [1, 3, 4]], '--', color=color)
    ax.plot([0,0,0], VP[0, [4, 5, 6]], VP[1, [4, 5, 6]], '--o', color=color)


def get_virtual(HP):

    # Contact patch location
    CPy = HP[1,-1]
    CPz = HP[2,-1]

    # Interpolate inboard wishbone points to wheel center plane
    IUy = HP[1, 0] - HP[0, 0] * (HP[1, 0] - HP[1, 1]) / (HP[0, 0] - HP[0, 1])
    IUz = HP[2, 0] - HP[0, 0] * (HP[2, 0] - HP[2, 1]) / (HP[0, 0] - HP[0, 1])

    ILy = HP[1, 3] - HP[0, 3] * (HP[1, 3] - HP[1, 4]) / (HP[0, 3] - HP[0, 4])
    ILz = HP[2, 3] - HP[0, 3] * (HP[2, 3] - HP[2, 4]) / (HP[0, 3] - HP[0, 4])

    # Instant Center
    m1 = (HP[2, 2] - IUz) / (HP[1, 2] - IUy)  # slope of upper wishbone line
    m2 = (HP[2, 5] - ILz) / (HP[1, 5] - ILy)  # slope of lower wishbone line
    b1 = HP[2, 2] - m1 * HP[1, 2]  # intercept of upper wishbone line
    b2 = HP[2, 5] - m2 * HP[1, 5]  # intercept of lower wishbone line
    ICy = (b2 - b1) / (m1 + m2)  # instant center y coordinate
    ICz = m1 * ICy + b1  # instant center x coordinate

    # Roll Center
    RCy = 0  # assuming symmetric suspension geometry
    if math.isfinite(ICy):
        m = (CPz - ICz) / (CPy - ICy)  # slope of the n-line
        RCz = CPz - m * CPy  # roll center height (z coordinate)
    else:
        RCz = 0

    VP = np.array([[IUy, IUz],
                   [ILy, ILz],
                   HP[1:3, 2],
                   HP[1:3, 5],
                   [ICy, ICz],
                   [RCy, RCz],
                   [CPy, CPz]]).transpose()
    return VP
